### DISCUSSION TITLE
[BUG]一个逗号引发的BUG

### DISCUSSION THREAD

**Discussion by BunnySakura** [adityatelange/hugo-PaperMod#1030] [Q&A] [answered]
# English：
Google Search Console alerted me to a problem with my blog: 

**Structured data that can not be parsed, The value type is incorrect**

Then I use the [web tool](https://validator.schema.org/ "web tool") to detect errors. 

example site: 
```
https://333rd.net/posts/read/%E7%BB%93%E6%9E%84%E4%BD%93%E7%9A%84%E4%BA%8C%E4%B8%89%E4%BA%8B/
```

The error was caused by a comma `,`, And I found the code that generated it  in **37** lines of **layous/partials/templates/schema_json.html**.

The culprit as follows:
```
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [, <--------THE comma `,`。
    {
```

The code that generated it is in:
```
    {{- if (and $index)}}, {{end }}
```
Hope to repair, thank you!

# 中文：
Google Search Console提示我网站存在**无法解析的结构化数据-值类型不正确**错误。

用[网页工具](https://validator.schema.org/ "网页工具")检测出来错误源于一个多出来的`,`逗号。如下：
```
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [, <--------罪魁祸首的逗号。
    {
```

我找到了生成它的代码在**layous/partials/templates/schema_json.html**的**37**行。
```
    {{- if (and $index)}}, {{end }}
```
希望修复，谢谢！

**Comment by BunnySakura** [comment]
**临时修复方案**：
```python
import os


def get_all_files(dir):
    """使用listdir循环遍历"""
    if not os.path.isdir(dir):
        print(dir)
        return
    dirlist = os.walk(dir)
    for root, dirs, files in dirlist:
        for file in files:
            yield os.path.join(root, file)


path = "./public"
for file in get_all_files(path):
    if file.split('.')[-1] == "html":
        with open(file, "r+") as f:
            text = f.readlines()
            for line in range(len(text)):
                if "  \"itemListElement\": [, " in text[line]:
                    text[line] = "  \"itemListElement\": [\n"
                    f.seek(0, 0)
                    f.writelines(text)
                    f.flush()
                    print(file, "is OK!")
                    break

```
**使用方法：**
新建`fix.py`，放在`public`同目录下，运行：
```bash
python3 fix.py
```

**Comment by smailzhu** [comment]
Could you share the minimal working example repository to reproduce  the issue?

**Reply by BunnySakura** [reply]
- English
  **First of all, thank you for your help!**

  I used the script above to alleviate the problem, which is no longer present on my site. But the problem persists, and it seems to be only in the most recent version of PaperMode. Here's an example from the site in disscussion: #1021:

  https://er4hn.info/blog/2022.08.05-djb_vs_usgov/

  Using Google's web detection tool, [web tools](https://validator.schema.org/ "Web tools") , you can get a sense of what we're talking about.

- 中文
  **首先感谢你的帮助！**

  我使用上面的脚本缓解了这个问题，在我的网站现在已经没有这个状况。但是问题依旧存在，而且似乎在最近的PaperMode版本才存在这个问题。此处以这个讨论中的网站举例：#1021：

  https://er4hn.info/blog/2022.08.05-djb_vs_usgov/

  使用谷歌网页检测工具：[网页工具](https://validator.schema.org/ "网页工具")，你就能得到我们讨论的问题所在。

**Reply by smailzhu** [reply]
Could you provide **minimal working example** for me to reproduce the problem?

Including hugo version and papermod version.

**Reply by BunnySakura** [reply]
- English
  Hello, I uploaded [a minimal environment](https://drive.google.com/file/d/1t7RgFoTO2YgvU6YaY8EXxy-WARWpQ09Z/view?usp=sharing) to reproduce the problem, just need to run the latest version of Hugo to generate the problem site. 

  I also noticed that this problem seems to be related to `config.yml`. YML 'configuration, but I'm not sure which one.

- 中文
  你好，我上传了一个复现问题的[最小环境](https://drive.google.com/file/d/1t7RgFoTO2YgvU6YaY8EXxy-WARWpQ09Z/view?usp=sharing)，只需要运行最新版的hugo就能生成问题网站。

  另外我发现这个问题好像和`config.yml`配置有关，但是我还不清楚具体和哪一项配置有关。

**Reply by smailzhu** [reply]
I cannot find the post you mentioned before. 

After I start the server with `hugo server` .
I can find only one page at `http://localhost:1313/posts/life/1st/`.

The `itemListElement` looks well...
![image](https://user-images.githubusercontent.com/9101328/188321149-ff06e83b-6d31-49bb-8640-829c9b8fa571.png)

How's the MWE works on your machine?

**Comment by adityatelange** [comment, marked as answer]
@BunnySakura Can you try using baseURL as
```yml
baseURL: https://website.com/                    # notice the trailing '/'
```

**Reply by adityatelange** [reply]
I'll add this fix to FAQs as well so that anyone else facing this issue will get instant resolution

**Reply by BunnySakura** [reply]
- English
  I think I found the problem and solved it. Yes, you're right.

  Parameters `baseURL` in `config.yml` must be set as you say:
  ```yaml
  baseURL: https://website.com/                    # notice the trailing '/'
  ```

  Previously I used the `hugo server` Command to create the site, and no problems occurred. But then I used the Nginx reverse proxy to create the web site, and there was a problem with using `hugo` to generate the HTML file.

  Finally Thank you for your help🤣! ! !

- 中文
  我想我发现并解决了问题。是的，你是对的。

  `config.yml`中，`baseURL`参数必须如你所说进行设置：
  ```yaml
  baseURL: https://website.com/                    # notice the trailing '/'
  ```

  之前我用`hugo server`命令创建网站，此时不会发生问题。但是之后我用Nginx反向代理来创建网站，此时使用`hugo`生成HTML文件就会出现问题。

  最后感谢你的帮助🤣！！！

**Comment by smailzhu** [comment]
I guess it may caused by some emoji or special character (maybe CJK)...
Because the position of the `itemListElement` should start from `1`.

**Reply by BunnySakura** [reply]
Problem solved, thanks for your help!
For more details: https://github.com/adityatelange/hugo-PaperMod/discussions/1030#discussioncomment-3549089

问题已经解决了，感谢你的帮助！
详情参考：https://github.com/adityatelange/hugo-PaperMod/discussions/1030#discussioncomment-3549089
